internal/k8s/cache/gateway.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type gatewayReconciler struct {
+	client       client.Client
+	eventHandler cache.ResourceEventHandler
+	logrus.FieldLogger
+}
+
+// NewGatewayController creates the gateway controller from mgr. The controller will be pre-configured
+// to watch for Gateway objects across all namespaces.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+	r := &gatewayReconciler{
+		client:       mgr.GetClient(),
+		eventHandler: eventHandler,
+		FieldLogger:  log,
+	}
+	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	// Fetch the Gateway from the cache.
+	gateway := &gatewayapi_v1alpha1.Gateway{}
+	err := r.client.Get(ctx, request.NamespacedName, gateway)
+	if errors.IsNotFound(err) {
+		r.Error(nil, "Could not find Gateway %q in Namespace %q", request.Name, request.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check if object is deleted.
+	if !gateway.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.eventHandler.OnDelete(gateway)
+		return reconcile.Result{}, nil
+	}
+
+	// Pass the new changed object off to the eventHandler.
+	r.eventHandler.OnAdd(gateway)
```

Comment:
I guess `OnAdd` makes sense here since, regardless of whether this was actually a create or an update, this will have the effect of overwriting the object in the Contour cache. Maybe a brief comment to explain why `OnAdd` is fine whether it's an add or an update would be helpful.